ctrimages/imageManager.go

```diff
@@ -0,0 +1,143 @@
+// Package ctrimages provides an image manager that manages and caches container images.
+package ctrimages
+
+import (
+	"context"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/remotes/docker"
+	log "github.com/sirupsen/logrus"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+)
+
+// ImageState is used to synchronize image pulling to avoid pulling the same image multiple times concurrently.
+type ImageState struct {
+	sync.Mutex
+	pulled bool
+}
+
+// NewImageState creates a new ImageState object that can be used to synchronize image pulling.
+func NewImageState() *ImageState {
+	state := new(ImageState)
+	state.pulled = false
+	return state
+}
+
+// ImageManager manages the images that have been pulled to the node.
+type ImageManager struct {
+	sync.Mutex
+	snapshotter  string						 // image snapshotter
+	cachedImages map[string]containerd.Image // Cached container images
+	imageStates  map[string]*ImageState
+	client       *containerd.Client
+}
+
+// NewImageManager creates a new imagemanager that can be used to fetch container images.
+func NewImageManager(client *containerd.Client, snapshotter string) *ImageManager {
+	log.Info("Creating image manager")
+	manager := new(ImageManager)
+	manager.snapshotter = snapshotter
+	manager.cachedImages = make(map[string]containerd.Image)
+	manager.imageStates = make(map[string]*ImageState)
+	manager.client = client
+	return manager
+}
+
+// pullImage fetches an image and adds it to the cached image list
+func (mgr *ImageManager) pullImage(ctx context.Context, imageName string) error {
+	var err error
+	var image containerd.Image
+
+	imageURL := getImageURL(imageName)
+	local, _ := isLocalDomain(imageURL)
+	if local {
+		// Pull local image using HTTP
+		resolver := docker.NewResolver(docker.ResolverOptions{
+			Client: http.DefaultClient,
+			Hosts: docker.ConfigureDefaultRegistries(
+				docker.WithPlainHTTP(docker.MatchAllHosts),
+			),
+		})
+		image, err = mgr.client.Pull(ctx, imageURL,
+			containerd.WithPullUnpack,
+			containerd.WithPullSnapshotter(mgr.snapshotter),
+			containerd.WithResolver(resolver),
+		)
+	} else {
+		// Pull remote image
+		image, err = mgr.client.Pull(ctx, imageURL,
+			containerd.WithPullUnpack,
+			containerd.WithPullSnapshotter(mgr.snapshotter),
+		)
+	}
+	if err != nil {
+		return err
+	}
+	mgr.Lock()
```

Comment:
that's a bug. you need to protect the whole function body with a lock.

Reply:
Not sure if I see the bug, the imagemanager mutex is only used to synchronise access to the imagemanager internal maps. There then also is an additional lock for each image to synchronize the pulling of that specific image. Although I agree a test would be good to have.

Reply:
please go over all the comments, I think the concurrency mechanism needs to be re-visited for more simplicity and testing, of course